pycbc/inference/models/marginalized_gaussian_noise.py

```diff
@@ -325,3 +326,293 @@ def _loglr(self):
+        setattr(self._current_stats, 'maxl_phase', numpy.nan)
+        return -numpy.inf
+
+    def _loglr(self, return_unmarginalized=False):
+        r"""Computes the log likelihood ratio,
+
+        .. math::
+
+            \log \mathcal{L}(\Theta) = \sum_i
+                \left<h_i(\Theta)|d_i\right> -
+                \frac{1}{2}\left<h_i(\Theta)|h_i(\Theta)\right>,
+
+        at the current parameter values :math:`\Theta`.
+
+        Returns
+        -------
+        float
+            The value of the log likelihood ratio.
+        """
+        params = self.current_params
+        try:
+            wfs = self.waveform_generator.generate(**params)
+        except NoWaveformError:
+            return self._nowaveform_loglr()
+        except FailedWaveformError as e:
+            if self.ignore_failed_waveforms:
+                return self._nowaveform_loglr()
+            else:
+                raise e
+
+        # ---------------------------------------------------------------------
+        # Some optimizations not yet taken:
+        # * higher m calculations could have a lot of redundancy
+        # * fp/fc need not be calculated except where polarization is different
+        # * may be possible to simplify this by making smarter use of real/imag
+        # ---------------------------------------------------------------------
+        lr = 0.
+        hds = {}
+        hhs = {}
+        for det, modes in wfs.items():
+            if det not in self.dets:
+                self.dets[det] = Detector(det)
+
+            fp, fc = self.dets[det].antenna_pattern(self.current_params['ra'],
+                                                    self.current_params['dec'],
+                                                    self.pol,
+                                                    self.current_params['tc'])
+
+            # loop over modes and prepare the waveform modes
+            # we will sum up zetalm = glm <ulm, d> + i glm <vlm, d>
+            # over all common m so that we can apply the phase once
+            zetas = {}
+            rlms = {}
+            slms = {}
+            for mode in modes:
+                l, m = mode
+                ulm, vlm = modes[mode]
+
+                # whiten the waveforms
+                # the kmax of the waveforms may be different than internal kmax
+                kmax = min(max(len(ulm), len(vlm)), self._kmax[det])
+                slc = slice(self._kmin[det], kmax)
+                ulm[self._kmin[det]:kmax] *= self._weight[det][slc]
+                vlm[self._kmin[det]:kmax] *= self._weight[det][slc]
+
+                # the inner products
+                # <ulm, d>
+                ulmd = ulm[slc].inner(self._whitened_data[det][slc]).real
+                # <vlm, d>
+                vlmd = vlm[slc].inner(self._whitened_data[det][slc]).real
+
+                # add inclination, and pack into a complex number
+                import lal
```

Comment:
If we expect lal, I would move this to the top. If you want to consider lal optional, you should catch if lal is not present when initializing an instance of this class.

Reply:
@spxiwh do you know if that is worked out anywhere? I think you are probably right, just wondering if someone has written it down.